panel/panel.py

```diff
@@ -136,64 +122,56 @@ def run_update_panel_cmd(panel, text=None):
-    tmpl = " {LINENO:>5}:{START:<4} {ERR_TYPE:7} {linter:>12}: {code:12} {MSG}"
-    return tmpl.format(
-        LINENO=lineno, START=start, ERR_TYPE=error_type, MSG=msg, **dic)
+def format_row(item):
+    line = item["line"] + 1
+    start = item["start"] + 1
+    tmpl = " {LINE:>5}:{START:<4} {error_type:7} {linter:>12}: {code:12} {msg}"
+    return tmpl.format(LINE=line, START=start, **item)
 
 
-def filter_ok(check_val, key, panel_filter):
+def passes_listing(error, panel_filter):
     """Check value against white- and blacklisting and return bool."""
     # check whitelisting
     if not panel_filter:
         return True
 
-    if key in panel_filter and check_val not in panel_filter[key]:
-        return False
-
-    # check blacklisting
-    exclude_key = "exclude_" + key
-    if exclude_key in panel_filter and check_val in panel_filter[exclude_key]:
-        return False
-
+    COMMAND_ARG_TO_ERROR_KEY = {
+        "types": "error_type",
+        "codes": "code",
+        "linter": "linter"
+    }
+
+    for key, check_val in COMMAND_ARG_TO_ERROR_KEY.items():
+        # check whitelisting
+        if key in panel_filter:
+            if error[check_val] not in panel_filter[key]:
+                return False
+
+        # check blacklisting
+        exclude_key = "exclude_" + key
+        if exclude_key in panel_filter:
+            if error[check_val] in panel_filter[exclude_key]:
+                return False
+
+    # if all checks passed return True
     return True
 
 
-def get_view_lines(view_dict, panel_filter):
-    view_lines = []
-    for lineno, line_dict in sorted(view_dict["line_dicts"].items()):
-        for error_type in WARN_ERR:
-            if not filter_ok(error_type, "types", panel_filter):
-                continue
-
-            err_dict = line_dict.get(error_type)
-            if not err_dict:
-                continue
-            items = sorted(err_dict, key=lambda k: k['start'])
-
-            items = [
-                item for item in items
-                if filter_ok(item['linter'], "linter", panel_filter) and
-                filter_ok(item['code'], "codes", panel_filter)
-            ]
-            for item in items:
-                view_lines.append((lineno, error_type, item))
-
-    return view_lines
+def get_buf_lines(buf_errors, panel_filter):
+    """Filter raw_errors through white- and blacklisting, then sort them."""
+    buf_errors = [e for e in buf_errors if passes_listing(e, panel_filter)]
+    return sorted(buf_errors, key=lambda x: (x["line"], x["start"], x["end"]))
 
 
 def fill_panel(window, update=False, **panel_filter):
-    errors = persist.errors.data
+
+    errors = persist.raw_errors
     if not errors:
         return
 
-    errors = filter_errors(window, errors)
-    errors = dedupe_views(errors)
+    errors = get_window_raw_errors(window, errors)
 
-    path_dict, base_dir = create_path_dict(errors)
+    path_dict, base_dir = create_path_dict(window, errors)
```

Comment:
Here we have not errors but errors_by_bid. errors means a list, errors_by_bid relates naturally to a dict. In create_path_dict you call this arg bids, which is simply not what it is. It is enough though to pass down only the `.keys()`, bc the dict itself is never used.